src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,489 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <sys/wait.h>
+
+#include "hexdecoct.h"
+#include "parse-util.h"
+#include "sd-daemon.h"
+#include "sd-messages.h"
+
+#include "bpf-dlopen.h"
+#include "bpfd-conf.h"
+#include "bpfd-manager.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "fs-util.h"
+#include "hashmap.h"
+#include "missing_syscall.h"
+#include "mkdir.h"
+#include "process-util.h"
+#include "signal-util.h"
+#include "socket-util.h"
+#include "stdio-util.h"
+#include "umask-util.h"
+
+static void btf_free(struct btf **btf) {
+        assert(btf);
+
+        sym_btf__free(*btf);
+}
+
+static void toggle_bpf_stats_enabled(bool enable) {
+        int r;
+
+        const char *value = enable ? "1" : "0";
+
+        r = write_string_file("/proc/sys/kernel/bpf_stats_enabled", value, WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (r < 0)
+                log_error_errno(r, "Failed to write '%c' to /proc/sys/kernel/bpf_stats_enabled, ignoring: %m", *value);
+}
+
+static int get_memlock_by_fd(int fd, uint64_t *ret_memlock) {
+        _cleanup_free_ char *fdinfo = NULL;
+        char *memlock_line;
+        char path[STRLEN("/proc/self/fdinfo/") + DECIMAL_STR_MAX(int)];
+        uint64_t v;
+        int r;
+
+        xsprintf(path, "/proc/self/fdinfo/%i", fd);
+        r = read_full_virtual_file(path, &fdinfo, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read memlock from /proc/self/fdinfo/%i, ignoring: %m", fd);
+
+        memlock_line = find_line_startswith(fdinfo, "memlock:");
+        if (!memlock_line)
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Failed to read memlock from /proc/self/fdinfo/%i, ignoring: %m", fd);
+
+        memlock_line += strspn(memlock_line, WHITESPACE);
+        memlock_line[strcspn(memlock_line, WHITESPACE)] = 0;
+
+        r = safe_atou64(memlock_line, &v);
+        if (r < 0)
+                return log_error_errno(r, "Failed to convert memlock string to unsigned integer, ignoring: %m");
+
+        *ret_memlock = v;
+        return 0;
+}
+
+static int fetch_bpf_maps_and_progs(Varlink *link) {
+        _cleanup_hashmap_free_ Hashmap *map_name_by_id = NULL;
+        uint32_t id = 0;
+        JsonVariant *prog_and_map_list = NULL;
+        int r;
+
+        map_name_by_id = hashmap_new(NULL);
+        if (!map_name_by_id)
+                return log_oom();
+
+        /* Iterate through all BPF maps to get map names, map types, and memlock */
+        while (sym_bpf_map_get_next_id(id, &id) >= 0) {
+                struct bpf_map_info map_info = {};
+                uint32_t map_info_len = sizeof(map_info);
+                _cleanup_close_ int fd = -EBADF;
+                uint64_t memlock = 0;
+                char *map_name = NULL;
+                JsonVariant *v;
```

Comment:
the reference is leaked. You need to make this `_cleanup_(json_varian_unrefp) JsonVariant *v = NULL;`